src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/PcsVmrBackFlower.cs

```diff
@@ -0,0 +1,151 @@
+            ICoherencyUpdateResolver coherencyUpdateResolver,
+            IAssetLocationResolver assetLocationResolver,
+            IFileSystem fileSystem,
+            ILogger<VmrCodeFlower> logger)
+        : base(vmrInfo, sourceManifest, dependencyTracker, dependencyFileManager, vmrCloneManager, repositoryCloneManager, localGitClient, localGitRepoFactory, versionDetailsParser, vmrPatchHandler, workBranchFactory, basicBarClient, libGit2Client, coherencyUpdateResolver, assetLocationResolver, fileSystem, logger)
+    {
+        _sourceManifest = sourceManifest;
+        _dependencyTracker = dependencyTracker;
+        _vmrCloneManager = vmrCloneManager;
+        _repositoryCloneManager = repositoryCloneManager;
+    }
+
+    public async Task<(bool HadUpdates, NativePath RepoPath)> FlowBackAsync(
+        string mappingName,
+        Build build,
+        string baseBranch,
+        string targetBranch,
+        CancellationToken cancellationToken = default)
+    {
+        (SourceMapping mapping, ILocalGitRepo targetRepo) = await PrepareVmrAndRepo(
+            mappingName,
+            build,
+            baseBranch,
+            targetBranch,
+            cancellationToken);
+
+        Codeflow lastFlow = await GetLastFlowAsync(mapping, targetRepo, currentIsBackflow: true);
+
+        var hadUpdates = await FlowBackAsync(
+            mapping,
+            targetRepo,
+            lastFlow,
+            build.Commit,
+            build,
+            baseBranch,
+            targetBranch,
+            true,
+            cancellationToken);
+
+        return (hadUpdates, targetRepo.Path);
+    }
+
+    private async Task<(SourceMapping, ILocalGitRepo)> PrepareVmrAndRepo(
+        string mappingName,
+        Build build,
+        string baseBranch,
+        string targetBranch,
+        CancellationToken cancellationToken)
+    {
+        // Prepare the VMR
+        await _vmrCloneManager.PrepareVmrAsync(
+            [build.GetRepository()],
+            [build.Commit],
+            build.Commit,
+            cancellationToken);
+
+        // Prepare repo
+        SourceMapping mapping = _dependencyTracker.GetMapping(mappingName);
+        var remotes = new[] { mapping.DefaultRemote, _sourceManifest.GetRepoVersion(mapping.Name).RemoteUri }
+            .Distinct()
+            .OrderRemotesByLocalPublicOther()
+            .ToList();
+
+        ILocalGitRepo targetRepo;
+
+        // Now try to see if the target branch exists already
+        try
+        {
+            targetRepo = await _repositoryCloneManager.PrepareCloneAsync(
+                mapping,
+                remotes,
+                [baseBranch, targetBranch],
+                targetBranch,
+                cancellationToken);
+        }
+        catch (NotFoundException)
+        {
+            // If target branch does not exist, we create it off of the base branch
+            targetRepo = await _repositoryCloneManager.PrepareCloneAsync(
```

Comment:
is it possible that the base branch also doesn't exist?

Reply:
It should not be possible because it's either set to subscription's target branch in PCS or to current branch in `darc`.
If it happens, we definitely want to throw an exception and not catch it though as it should not be possible.